core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -620,6 +620,14 @@ private[spark] class TaskSetManager(
     // Note: "result.value()" only deserializes the value when it's called at the first time, so
     // here "result.value()" just returns the value and won't block other threads.
     sched.dagScheduler.taskEnded(tasks(index), Success, result.value(), result.accumUpdates, info)
+    // Kill other task attempts if any as the one attempt succeeded
+    for (attemptInfo <- taskAttempts(index) if attemptInfo.attemptNumber != info.attemptNumber
```

Comment:
I've only taken a quick look at this pr but I think there is a race here that if the speculative task finishes before this kill event can actually go out it will still be marked as success.  Please correct me if I'm wrong.

SparkHadoopMapRedUtil.commitTask can just log if something is already committed, (sometimes it gets CommitDeniedException, depending on timing).    At one point I had played with changing that to commit denied but other issues happened and I didn't have time to finish investigating.